modules/ocf_backups/files/upload-to-box

```diff
@@ -0,0 +1,444 @@
+        conn.setopt(conn.NOPROGRESS, False)
+
+        conn.file_bytes = os.path.getsize(file_path)
+        conn.fp = open(file_path, 'rb')
+
+        conn.setopt(conn.INFILESIZE, conn.file_bytes)
+        conn.setopt(conn.READFUNCTION, conn.fp.read)
+        conn.setopt(conn.UPLOAD, True)
+        conn.setopt(conn.URL, '{}{}/{}'.format(BOX_DAV, FOLDER, self.filename))
+        conn.setopt(conn.USERPWD, self.auth)
+
+    def progress(self, download_t, download_d, upload_t, upload_d):
+        """Store the upload progress percentage for periodic printing.
+
+        Passed to pycurl as a progress function for each upload
+        """
+        self.bytes_uploaded = upload_d
+        if upload_t > 0:
+            self.percentage = 100 * (upload_d / upload_t)
+
+    def get_progress(self, terminal_width):
+        """Get a nice progress bar including the file and percent complete
+
+        The progress bar has a dynamic width, so it must be constructed by
+        figuring out the correct width based on other data shown
+        """
+        first_half = '{} ['.format(self.filename)
+        second_half = '] {:.2f}%'.format(self.percentage)
+        width = terminal_width - len(first_half) - len(second_half)
+        filled = int(width * self.percentage / 100)
+
+        return '{}{}{}{}'.format(
+            first_half,
+            filled * '#',
+            (width - filled) * ' ',
+            second_half
+        )
+
+
+def box_api_call(url, data='', headers=[], method=''):
+    """Make an API call to Box.com (or anywhere really)"""
+    conn = pycurl.Curl()
+
+    response_data = BytesIO()
+    conn.setopt(conn.URL, url)
+    conn.setopt(conn.WRITEDATA, response_data)
+    conn.setopt(conn.HTTPHEADER, headers)
+
+    if data:
+        conn.setopt(conn.POSTFIELDS, data)
+
+    if method:
+        conn.setopt(conn.CUSTOMREQUEST, method)
+
+    conn.perform()
+    http_status = conn.getinfo(pycurl.HTTP_CODE)
+    conn.close()
+
+    # Check if HTTP status is in 2XX (success of some sort)
+    if http_status in range(200, 299):
+        return json.loads(response_data.getvalue().decode('utf-8'))
+    else:
+        print('HTTP {} on {} with {}'.format(http_status, url, data))
+        sys.exit(1)
+
+
+def get_access_token(creds):
+    """Gets a new access token for the Box.com API"""
+    with open('/opt/share/backups/box-refresh-token', 'r+') as f:
+        refresh_token = f.read().strip()
+
+        # Get an access token and a new refresh token from the API
+        post_data = '&refresh_token={}&client_id={}&client_secret={}'.format(
+            refresh_token,
+            creds['api_client_id'],
+            creds['api_client_secret']
+        )
+        response = box_api_call('https://app.box.com/api/oauth2/token',
+                                data='grant_type=refresh_token' + post_data)
+
+        # Write the new refresh token to the file
+        f.seek(0)
+        f.write(response['refresh_token'])
+        f.truncate()
```

Comment:
Pretty sure I just took it from [here](http://stackoverflow.com/questions/2424000/read-and-overwrite-a-file-in-python), but thanks :P